modin/config/pubsub.py

```diff
@@ -13,9 +13,80 @@
 
 """Module houses ``Parameter`` class - base class for all configs."""
 
+import warnings
 from collections import defaultdict
 from enum import IntEnum
-from typing import Any, Callable, DefaultDict, NamedTuple, Optional, Tuple
+from typing import (
+    TYPE_CHECKING,
+    Any,
+    Callable,
+    DefaultDict,
+    NamedTuple,
+    Optional,
+    Tuple,
+    cast,
+)
+
+if TYPE_CHECKING:
+    from .envvars import EnvironmentVariable
+
+
+class DeprecationDescriptor:
+    """
+    Describe deprecated parameter.
+
+    Parameters
+    ----------
+    parameter : type[Parameter]
+        Deprecated parameter.
+    new_parameter : type[Parameter], optional
+        If there's a replacement parameter for the deprecated one, specify it here.
+    when_removed : str, optional
+        If known, the exact release when the deprecated parameter is planned to be removed.
+    """
+
+    _parameter: type["Parameter"]
+    _new_parameter: Optional[type["Parameter"]]
+    _when_removed: str
+
+    def __init__(
+        self,
+        parameter: type["Parameter"],
+        new_parameter: Optional[type["Parameter"]] = None,
+        when_removed: Optional[str] = None,
+    ):
+        self._parameter = parameter
+        self._new_parameter = new_parameter
+        self._when_removed = "a future" if when_removed is None else when_removed
+
+    def deprecation_message(self, use_envvar_names: bool = False) -> str:
+        """
+        Generate a message to be used in a warning raised when using the deprecated parameter.
+
+        Parameters
+        ----------
+        use_envvar_names : bool, default: False
+            Whether to use environment variable names in the warning. If ``True``, both
+            ``self._parameter`` and ``self._new_parameter`` have to be a type of ``EnvironmentVariable``.
+
+        Returns
+        -------
+        str
+        """
+        name = (
+            cast("EnvironmentVariable", self._parameter).varname
```

Comment:
Why do you then declare `_parameter` with `type["Parameter"]` and not `EnvironmentVariable` initially?

Reply:
Well, that was because the `_deprecation_descriptor` is defined in the base `Parameter` class, so I thought it's worth putting the same type to the descriptor for consistency, but now I realize that it doesn't make much sense as we don't have nor planning to have other parameter types rather than `EnvironmentVariables`. So changed the signatures as you proposed.

Reply:
I am not sure this is the right solution. `pubsub.py` file contains `Parameter` class, which is a base for all configs, and it shouldn't depend on child classes. Same is true for DeprecationDescriptor, which can be expanded to all types of configs, but not just EnvVar.

Reply:
@YarShev So what's your proposal then? I see three options:

1. Change the signature of `DeprecationDescriptor` to accepting general `Parameter` type with an optional cast to `EnvVariable` on `generate_warning(use_envvar_names=True)`.
2. Leave `DeprecationDescriptor`'s signature as is and move it to the `envvars.py` (note, that the base `Parameter` class will still have a reference to the descriptor)
3. Leave everything as it is.

Personally, I'm voting for the 1st one, but won't complain if you pick any other option, as IMHO we're wasting too much time on this subtle detail

Reply:
1st is fine to me.